### PR TITLE
[WIP] Tail recursion warning and attribute

### PR DESCRIPTION
**Pull request by AviAvni**
Start working on https://github.com/fsharp/fslang-suggestions/issues/147

- [ ] add tests
- [ ] check for async { .. } and seq { .. }

image

### CODE REVIEW

**Comment by vasily-kirichenko** [comment]
`TailRecursionAttribute` is too long for my taste. What about `TailrecAttribute`? 
Scala uses `@tailrec` attribute.

**Comment by abelbraaksma** [comment]
This is a wonderful addition to F#! Writing either `[<TailRec>]` or `[<TailRecursion>]` wouldn't matter much to me (@vasily-kirichenko). In fact, I think if at all possible, `let rec f` vs `let tail rec f` (or `let tailrec f`) would have my preference, but that'd mean introducing a new keyword.

We may need a keyword if we are to allow this feature on local functions:

```
let factorial n =
    let rec loop i acc =   // this is tail-recursive, I think, but if it isn't I'd like the new warning
        match i with
        | 0 | 1 -> acc
        | _ -> loop (i-1) (acc * i)
    loop n 1
```

**Comment by forki** [comment]
Mhm since many many rec functions are local I think the introduction of a new tailrec keyword makes a lot of sense. If we decide on attributes then we should go with "TailRecursionAttribute". This is very similar to the existing RequireQualifiedAccessAttribute". 

Regarding the error message. I think we should make it really verbose and explain as much as possible.

**Comment by forki** [comment]
Forgot to mention that this is just wonderful. One of the best additions to the compiler in quite some time. Kudos for tackling this.

**Comment by forki** [comment]
There is also the possibility of taking a more radical approach and to always check for tail recursion. This has a couple of advantages. 
1) we don't need to introduce new syntax constructs nor attributes 
2) existing code will be checked without the need of change 
3) people can always opt-out by nowarn directive - that's already built in 
4) using tail recursion is whenever it is possible also the  best way to do things and should therefore the default choice of the language 
5) it is still non-breaking (ignoring "warning as error" mode like we usually do) 

Tbh I think this should be really considered. @dsyme what do you think?

**Comment by thinkbeforecoding** [comment]
@forki suggestion to check tail recursion systematically seems a good idea. 
Recursion is to my experience really often used in local nested functions. So it'd be a problem if attributes are limited to top level functions (didn't check).
There's a tailcall reserved keyword, but it seemed @dsyme didn't recommend its usage.

**Comment by vilinski** [comment]
+1 to check always

**Comment by AviAvni** [comment]
Thanks for everyone I will wait for @dsyme to decide about the details

**Comment by abelbraaksma** [comment]
I'm in favor of checking always, with one caveat: beginning users will be stymied by this warning, it requires a working knowledge of too many concepts to even understand the gist of what it means, let alone knowing how to fix it.

Either we should make it a L4 warning (are levels currently used?) or we should make sure that the warning includes a "ignore me" statement, for instance:

> FS8888: This function is marked recursive, but is not tail-recursive. If the recursion is deep or performance is of the essence, consider making it tail-recursive, otherwise, ignore this warning with #nowarn 8888.

Either way, I would prefer this as an option for experienced programmers. In 9 out of 10 cases it doesn't matter, as recursion won't go too deep anyway and performance isn't critical. If L3 warnings is the default, making it an L4 warning would suffice (though I am uncertain whether or not the levels are actually adhered to, or used, see for instance [this uservoice post](https://visualstudio.uservoice.com/forums/121579-visual-studio-ide/suggestions/6013742-enable-warnon-1182-unused-variables-in-f-warn), which wasn't converted?)

Warnings like these should either be leveled (L4, L5), or be allowed to be switched on/off on a more granular level than per-file (but that's OT for this issue, I know).

Even better yet: if someone feels like it, this could be switched on/of on the property page of a project (an option "analyze for tail-recursion").

**Comment by vladima** [comment]
Idea of detecting calls in non-tail position definitely sounds useful, however I think that implementation should be different. Reason: currently it is done at the codegen phase which is IDE never does (because  there is no point to generate actual binary on every keystroke) - so this error will never be reported in editor which to me significantly decreases its value.

**Comment by forki** [comment]
> In 9 out of 10 cases it doesn't matter, as recursion won't go too deep anyway

My (completely made up) statistics suggest the exact opposite. ;-)

**Comment by abelbraaksma** [comment]
@forki, I came to this idea from Brian on SO suggesting, esp. too beginners or moderately advanced programmers to stay away from even trying to use tail recursion: http://stackoverflow.com/a/1797302/111575.

From my own experience: looking at the my largest F# project I see some 50 or so `let rec` statements. Only a handful of which would benefit from this. That is not to say, I also have a few while-statements which I didn't dare to make recursive for fear of getting it wrong... (shame on me, I know, kudos to this fix, it can be a life saver!)

Thinking of which: I also encountered a few mutual recursive members. Would these be detected as tailcall with this fix?

**Comment by forki** [comment]
I absolutely agree with the "prefer higher order functions" part, but that
is actually not really related to this here.

Am 10.12.2016 6:20 nachm. schrieb "Abel Braaksma" <notifications@github.com

**Comment by abelbraaksma** [comment]
My train of thought was: if the tendency is to suggest beginning users to stay away from recursive, and if they don't, then stay away from tailcalls (makes sense, they can be hard, even for experienced programmers), then I think we ought to make sure this warning either doesn't throw them off, helps them, or is only available as an extra option for advanced users.

**Comment by forki** [comment]
stay away from tail-calls is a bad advice. really

**Comment by abelbraaksma** [comment]
> stay away from tail-calls is a bad advice. really

I'm just trying to follow consensus here. Saying "stay away" is perhaps too strong, but my point is simply: don't overwhelm beginning users with trying to explain what TCO is by means of warnings. We are trying to get more users, not less.

Regardless of your opinion on my ideas of tutoriability of warnings, this warning is a very valuable addition, just don't get my comments the wrong way, I am not arguing against implementing it, or against TCO in general (apologies if it looked that way).

**Comment by theimowski** [comment]
We already have the `rec` keyword for marking recursive functions, so if we provide yet another construct for marking tail recursion (be it an attribute, or special keyword), it might turn out tedious to use and confusing for new-comers. That said, I'm for making tail recursion warnings turned on by default.

**Comment by TeaDrivenDev** [comment]
I'm for always checking as well, without any special keywords or decorations. Recursion is so important in functional programming that one should be able to consider it "safe" by default, so the compiler telling you when it isn't is a good thing.
As for the beginners, I think that would actually help them too. Myself, I still don't always write proper tail calls, because it often doesn't matter, so when I *have* to, I sometimes don't really know how to get there. If I always had to do them properly to satisfy the compiler, I'd have much more practice with that.

**Comment by OnurGumus** [comment]
I would vote for a tailrec keyword. Always checking would yield unnecessary warnings.

**Comment by vasily-kirichenko** [comment]
I'm for `tailrec` keyword, too. It looks quite natural as a specific `rec`, it's far tidier that `[<TailRecursive>]`.

**Comment by cartermp** [comment]
Is the suggestion here to have `tailrec` be a replacement for `rec`? e.g.

```fsharp
let tailrec foo() = ...
```

**Comment by forki** [comment]
That is one of three proposed versions.

Am 12.12.2016 16:30 schrieb "Phillip Carter" <notifications@github.com>:

**Comment by vasily-kirichenko** [comment]
> Is the suggestion here to have tailrec be a replacement for rec? 

Yes.

**Comment by 0x53A** [comment]
this might be ambiguous, today I can declare a function named tailrec which takes a single parameter foo like

``` F#
let tailrec foo = ...
```

Since rec is already a keyword, and can not be used as a parameter name, I would propose

``` F#

let tail rec foo = ...

```

because that code is currently illegal.

Another option would be to use an attribute. Since the attribute does not need to be persisted to CIL, it could even be attached to local functions (which is not allowed today, but this would need to be changed).

**Comment by isaacabraham** [comment]
Don't want tail recursion checking + warning to be mandatory - at least, not as a first release. I suggest to make it optional via `tailrec` keyword (or attribute I suppose). We can always make `rec` and `tailrec` do the same thing in a future version *if* we wanted to.

**Comment by matthid** [comment]
If we add a new keyword/attribute I think the compiler should warn when the recursion is a tailcall but the new keyword/attribute is not used.

That said I personally don't like a new keyword.
What about the other way to always warn when it's not a tailcall and add a attribute to disable the warning '[<NoTailCall>]'.

**Comment by rojepp** [comment]
@matthid There is already a way to disable warnings `#nowarn`, we don't need another?

**Comment by OnurGumus** [comment]
@mathid. No, if you did not use tailrec there'd be no need for any warnings.

**Comment by matthid** [comment]
@OnurGumus my suggestion is basically the same as @forki suggested already https://github.com/Microsoft/visualfsharp/pull/1976#issuecomment-266203763

@rojepp because nowarn is a global approach to opt out of the feature, but you might want a straightforward way to opt out for a single instance...

**Comment by ncave** [comment]
Would adding a compiler option for it be better than a new attribute?

**Comment by AviAvni** [comment]
@dsyme please comment so I'll be able to continiue work on it

**Comment by abelbraaksma** [comment]
@AviAvni, on how it is implemented, is the warning shown _after compilation_, or during syntax checking while in the editor (i.e. as intellisense warning)?

**Comment by AviAvni** [comment]
@abelbraaksma for now its after compilation because the check is in Code Generation phase

**Comment by OnurGumus** [comment]
On the other hand I am thinking perhaps we should mark tail recursion from the call site. A function can recourse from two location and one can be tail reclusive and other is not
`let factorial n =
    let rec loop i acc =   // this is tail-recursive, but if it isn't I'd like the new warning
        match i with
        | 0 | 1 -> acc
        |2-> (loop (i-1) (acc * i) + loop (i-1) (acc * i))/2  //silly example but anyway, no tail rec here
        | _ -> loop (i-1) (acc * i) //tail rec here
    loop n 1`

**Comment by dsyme** [comment]
Some of my comments in the original design proposal are in the user voice comments: https://fslang.uservoice.com/forums/245727-f-language/suggestions/5663074-enable-a-compiler-warning-when-a-recursive-algorit

Here are some additional notes:

* @vladima is correct that the implementation really needs to be in the PostInferenceChecks or similar phase, which gets run in the IDE.  This is especially true for seq/async tailcalls.

* As noted in the UV comments, my preference is for an attribute.  I am generally against adding new keywords for this kind of thing.  There are various reasons, but one is that there are many other operational constraints which you could imagine having checked by the compiler (or addins to the compiler) - "UsesLimitedStack(1000)" or "AlwaysTerminates" or ...  

* Another is that the attribute can be applied to members, which can also be tail recursive, e.g.

```
type C() = 
    [<TailRec>]
    member x.M() = x.M()
```

* People mention "always checking" - I agree that if the attribute/keyword is used, then checking should report warnings or errors by default.  It should not, however, be checked if the attribute/keyword is not used.  It's not on by default when no attribute is used.

* Re "stay away from tail-calls is a bad advice. really" - I think Brian's advice in the SO thread was more nuanced - basically he's saying that in F# you need to write tail recursive functions less often, because the library (particularly List, Array and especially Seq) lets you avoid it very often.    But that's a pub discussion.

* Re "On the other hand I am thinking perhaps we should mark tail recursion from the call site" - this is certainly a concern.  I would really like to see both prototyped and we experiment with both to see which feels more natural and less intrusive.  It's possible that you could prototype the callsite version using a new library intrinsic ``val tailrec : 'T -> 'T`` used as ``tailrec (f x)``.

**Comment by AviAvni** [comment]
@dsyme thanks for your comment just ported you PR from codeplex now the warning is added when typing
image
I'll continue work on it on sunday

**Comment by AviAvni** [comment]
@dsyme first question about the async {} and seq {} now the compiler identify them as not tail call
image
Is there a way to  make them tail call?

second question in the gist of jack https://gist.github.com/jack-pappas/9860949
image
there is a mutual recursive function now the code in the pr identify that in line 17 that it's not tail call
but if i added in the end of the baz function more printfn lineit dosen't identify that line 23 is not a tail call
do we need to support this kind of mutual recursive function?

thanks

**Comment by dsyme** [comment]
> Is there a way to make them tail call?

These are not tailcalls in the usual .NET IL sense -  the use of bounded resources relies heavily on the implementation of the ``seq`` or ``async`` builder. It's an open question to determine what to do here.  I suppose it will have to be a separate PR, but it's a shame because I prefer if the feature doesn't bias F# programming away from using computation expressions in order to get the additional safety....

**Comment by dsyme** [comment]
> dosen't identify that line 23 is not a tail call... do we need to support this kind of mutual recursive function?

Yes, line 23 should be marked as not-a-tailcall

**Comment by AviAvni** [comment]
So to summarize
1.  I need to figure out why line 23 is not marked and fixed it
2. For this PR remove the checks in computation expression
3. Try to implement the tailrec intrinsic in this PR?

Thanks

**Comment by forki** [comment]
Tbh I think we need to check both in editor phase and in IL (or at least after last optimization).

**Comment by AviAvni** [comment]
@forki I can bring back the check in the code gen

**Comment by forki** [comment]
I think that would be good, not sure how others feel about it.

File: src/fsharp/FSComp.txt [hunk 1]
```diff
@@ -1332,3 +1332,4 @@ tcTupleStructMismatch,"One tuple type is a struct tuple, the other is a referenc
 3211,DefaultParameterValueNotAppropriateForArgument,"The default value does not have the same type as the argument. The DefaultParameterValue attribute and any Optional attribute will be ignored. Note: 'null' needs to be annotated with the correct type, e.g. 'DefaultParameterValue(null:obj)'."
 tcGlobalsSystemTypeNotFound,"The system type '%s' was required but no referenced system DLL contained this type"
 3213,typrelMemberHasMultiplePossibleDispatchSlots,"The member '%s' matches multiple overloads of the same method.\nPlease restrict it to one of the following:%s."
+3212,chkNotTailRecursive,"The member or function '%s' has the 'TailCallAttribute' attribute, but is not being used in a tail recursive way."
```
Comment (forki) [inline src/fsharp/FSComp.txt]: Maybe "was marked with"? Do we already have similar texts?

**Comment by dsyme** [comment]
> 3.	Try to implement the tailrec intrinsic in this PR?

Ideally yes.  This would be an F# 4.2 feature, and I think we would progress by first having one or two people try out the feature, e.g. try to use apply it to the F# codebase.  Having both the intrinsic and the attribute implemented would help.  

Note 

1. The presence of uses of the intrinsic in the TAST may easily mess up optimizations and/or codegen.  So removing it during the first optimization phase makes sense.    However this means you can't perform the check at code-gen time.

2. You will also need to decide if it appears in quotations - it shouldn't I think - so should again be removed during quotation processing.

3. Syntactically the use of an intrinsic has pros and cons.  For example, peope may expect you can pipeline it, e.g. ``let rec f x = ... tailrec (f x)`` becomes ``let rec f x = ...f x |>  tailrec``.  That won't work of course.  I can also see people definitely asking for a keyword here, so they don't need parentheses ``let rec f x = ... tailrec f x``

4. The current intended rule is "TailRecAttribute implies the function is only used in a tail recursive way within its recursive scope" .  Just to note that this is in tension with the various ways to increase the size of recursive scopes, e.g. ``namespace rec`` or ``module rec`` or even just a set of object-oriented type definitions.    Increasing the size of a recursive scope will mean spurious errors/warnings are given.  In this case, people will gravitatte to using ``tailrec``

Checking in codegen does make sense, to make sure the optimizer hasn't screwed things around.

**Comment by forki** [comment]
So intrinsics feature should be a separate PR?

**Comment by jack-pappas** [comment]
I've been thinking -- introducing a ``tailrec`` keyword or ``[<TailRecursive>]`` attribute doesn't seem strictly necessary. Unless there's a something I'm missing, it'd only serve to augment the analysis; for example, when the keyword/attribute is applied, the analysis pass might emit an error instead of a warning.

Given this is a rather complex piece of analysis with many edge cases to consider, what if we went the route of implementing it within something like VFPT for now? There, it could be enabled optionally while we work on getting the analysis correct; once it's ready, we'd roll the analysis code into the compiler itself and implement the keyword/attribute if it still seems useful. If VFPT isn't a good fit, what about implementing the analysis as a standalone tool (again, for now) on top of FCS -- we could provide it as a NuGet package which would install the tool into a project (so no manually configuration would be needed, and it'd still be easy to iterate quickly).

**Comment by KevinRansom** [comment]
Jack,

I think an IDE analyzer rather than language keyword is the way to go.

It could just kick out a warning that
                Let rec foo =
                                Blah …
                                Blah …

Is not be tail-recursive.  The il spec regards the .tail instruction as a hint to the jitand it is up to the implementor of the run-time to implement support or just ignore it.

I know jit-devs who believe and state quite loudly that if we have to use the .tail command to get the right behavior at all it is a jit bug.

So … in short try to think of ways not to bake this into the language, instead make it an IDE analyzer.  We may have to figure out ways to make it easy to switch on and off … but that is probably better than baking it in.


Thanks

Kevin

**Comment by forki** [comment]
Analyzer are roslyn based. This feature  is something that would useful in the whole ecosystem. We should really think about that and make the compiler smart instead of relying on analyzers. I'm not saying that we have to introduce a keyword or attribute - just that we should use analyzers only as last resort.

**Comment by KevinRansom** [comment]
The C# compiler runs analyzers.  We should make that work.

**Comment by forki** [comment]
I can already run as many analyzers as I want during my FAKE build. That's not the point.

**Comment by forki** [comment]
Btw it's nice that people say it's a jit bug if you have to emit tail calls. Are these bugs fixed?

**Comment by forki** [comment]
Even if the jitter could detect tail calls, then we still have to write it tail recursive. And the whole idea of this feature is to detect this somehow in the compiler.

**Comment by KevinRansom** [comment]
@forki I believe they would consider fixing the Jit to improve code generation where they believe tail recursion should occur and it isn't.  Provide a repro of where the jit does the wrong thing and they will improve it.

**Comment by dsyme** [comment]
@jack-pappas 

> Unless there's a something I'm missing, it'd only serve to augment the analysis; for example, when the keyword/attribute is applied, the analysis pass might emit an error instead of a warning.

There seems to be a belief going around that F# function calls are normally tail recursive and we can just warn if they aren't.  This isn't the case - a huge number of F# calls (the majority of syntactic call sites) are not tail recursive - even within the recursive scope of the function or member.

So some kind of attribute is needed to indicate that tail recursion is intended.  Otherwise we end up flagging almost everything.

**Comment by dsyme** [comment]
Everyone - his is in the context of [the approved language RFC FS-1011](https://github.com/fsharp/fslang-design/blob/master/RFCs/FS-1011-warn-on-recursive-without-tail-call.md).  We should probably put design feedback there.

@KevinRansom I'm not opposed to any particular analyzer of course, likewise I wouldn't be opposed to an FSharpLint rule.  (FSharpLint rules are active in Ionide by defaut).

Writing a code fix to make something tail recursive would be hard, as there are quite a few ways to do it (see Expert F# 4.0 chapter ~15)

**Comment by forki** [comment]
The thing is even with analyzer we need to commit that decision with the
code or we will flag everything.

Am 18.01.2017 15:55 schrieb "Don Syme" <notifications@github.com>:

**Comment by jack-pappas** [comment]
@dsyme Agreed, we should move discussion about the design for this feature over to fsharp/fslang-design#82.

**Comment by jack-pappas** [comment]
@dsyme I also realized after your comment today that we may be talking about two different things -- tail calls in general (which don't necessarily imply any recursion) and functions designed to be (mutually-)recursive (where any call sites to functions in that set of mutually-recursive functions should be tail-calls to avoid overflowing the stack). I think there's an easy way to get some low-hanging fruit on the mutually-recursive functions, and I've outlined the analysis algorithm in fsharp/fslang-design#82.

**Comment by KevinRansom** [comment]
@AviAvni @dsyme 

Thanks for this PR and lively discussion but it doesn't seem that we are anywhere close to having the necessary investigations and designs complete yet for a tail recursion feature.

I'm going to close this PR for now, the discussion needs to occur on user voice.  When consensus or an approach has been determined, then please feel free to resubmit this or an alternate PR.

Thank you,

Kevin

**Comment by forki** [comment]
@kevinransom please reopen. This is one of the most interesting pull requests in recent history

**Comment by KevinRansom** [comment]
The language design is 
https://github.com/fsharp/fslang-design/blob/master/RFCs/FS-1011-warn-on-recursive-without-tail-call.md

This interesting thread is not being captured at the right spot.

I will add a link from :
https://github.com/fsharp/fslang-design/blob/master/RFCs/FS-1011-warn-on-recursive-without-tail-call.md

back to this thread.

Please consider continuing the discussion at the RFC.

**Comment by dsyme** [comment]
@KevinRansom I think we could leave this open - any efforts to make progress on this particular issue are really valuable, even half-way.  It's a really tricky feature to get right, and the thing we need most is a series of prototypes for people to use.  A WIP PR is one way to get that (and get eyes on it)?

**Comment by realvictorprm** [comment]
Can the FSSF help with triggering this discussion again @dsyme?

**Pull request closed without merging**